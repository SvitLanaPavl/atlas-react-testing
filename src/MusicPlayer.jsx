import CurrentlyPlaying from "./components/CurrentlyPlaying";
import Playlist from "./components/Playlist";

export default function MusicPlayer() {
  return (
    <div className='mx-auto flex h-full w-full max-w-5xl flex-col overflow-hidden rounded-lg shadow-lg md:flex-row'>
      <CurrentlyPlaying />
      <Playlist />
    </div>
  );
}
