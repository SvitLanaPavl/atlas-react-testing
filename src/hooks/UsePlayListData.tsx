import { useEffect, useState } from 'react';

export interface Song {
  title: string;
  artist: string;
  duration: string;
  cover: string;
}

export const usePlaylistData = () => {
  const [data, setData] = useState<Song[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  useEffect(() => {
    const fetchPlaylist = async () => {
      setLoading(true);

      // Simulate a delay
      await new Promise((resolve) => setTimeout(resolve, 3000)); // 3-second delay

      fetch('https://raw.githubusercontent.com/atlas-jswank/atlas-music-player-api/main/playlist')
        .then((response) => response.json())
        .then((data: Song[]) => {
          setData(data);
          setLoading(false);
        })
        .catch((error) => {
          console.error('Error fetching playlist:', error);
          setLoading(false);
        });
    };

    fetchPlaylist();
  }, []);

  return { data, loading };
};
