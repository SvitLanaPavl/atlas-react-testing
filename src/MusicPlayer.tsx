import React, { useState } from 'react';
import CurrentlyPlaying from './components/CurrentlyPlaying';
import Playlist from './components/Playlist';
import { usePlaylistData } from './hooks/UsePlayListData';



const MusicPlayer: React.FC = () => {
  const {data: playlist, loading} = usePlaylistData();
  const [currentSongIndex, setCurrentSongIndex] = useState<number>(0);
  const [isShuffle, setIsShuffle] = useState<boolean>(false);
  const [songHistory, setSongHistory] = useState<number[]>([]); // Song history to track previous songs

  

  const handleSongChange = (index: number) => {
    if (index >= 0 && index < playlist.length) {
      setSongHistory((prevHistory) => [...prevHistory, currentSongIndex]);
      setCurrentSongIndex(index);
    }
  };

  const handleShuffleToggle = () => {
    setIsShuffle((prevShuffle) => !prevShuffle);
    setSongHistory([]); // Clear history when toggling shuffle
  };

  const handleNextSong = () => {
    if (isShuffle) {
      // Shuffle mode: pick a random song and store current song in history
      const randomIndex = Math.floor(Math.random() * playlist.length);
      setSongHistory((prevHistory) => [...prevHistory, currentSongIndex]);
      setCurrentSongIndex(randomIndex);
    } else {
        setCurrentSongIndex(currentSongIndex + 1);
      }
  };

  const handlePrevSong = () => {
    if (isShuffle && songHistory.length > 0) {
      // In shuffle mode, go back to the last song in history
      const lastSongIndex = songHistory[songHistory.length - 1];
      setSongHistory((prevHistory) => prevHistory.slice(0, -1)); // Remove last entry from history
      setCurrentSongIndex(lastSongIndex);
    } else {
      // In normal mode, go back one song
      setCurrentSongIndex(currentSongIndex - 1);
    }
  };

  return (
    <div className="mx-auto flex h-full w-full max-w-5xl flex-col overflow-hidden rounded-lg shadow-lg md:flex-row">
      {loading ? (
        <p>Loading playlist...</p>
      ) : (
        <>
          {playlist.length > 0 && (
            <CurrentlyPlaying
              song={playlist[currentSongIndex]}
              onPrev={handlePrevSong}
              onNext={handleNextSong}
              isFirstSong={currentSongIndex === 0}
              isLastSong={currentSongIndex === playlist.length - 1}
              onShuffleToggle={handleShuffleToggle}
              isShuffle={isShuffle}
            />
          )}
          <Playlist
            playlist={playlist}
            currentSongIndex={currentSongIndex}
            onSongSelect={handleSongChange}
          />
        </>
      )}
    </div>
  );
};

export default MusicPlayer;
