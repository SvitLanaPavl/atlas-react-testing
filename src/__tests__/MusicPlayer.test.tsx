import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import MusicPlayer from '../MusicPlayer'; // Adjust the path to your component
import { usePlaylistData } from '../hooks/UsePlayListData'; // Mock the hook

// Mock the usePlaylistData hook
vi.mock('../hooks/UsePlayListData');

describe('MusicPlayer Component', () => {

  it('displays playlist and currently playing song once loaded', async () => {
    // Mock the hook to return playlist data
    (usePlaylistData as vi.Mock).mockReturnValue({
      data: [
        { id: 1, title: 'Painted in Blue', artist: 'Soul Canvas', duration: '5:55' },
        { id: 2, title: 'Tidal Drift', artist: 'Echoes of the Sea', duration: '8:02' },
      ],
      loading: false,
    });

    render(<MusicPlayer />);

    // Ensure the playlist and song are rendered
    expect(await screen.findByText('Painted in Blue')).toBeInTheDocument();
    expect(screen.getByText('Soul Canvas')).toBeInTheDocument();
  });

  it('handles song navigation (next and previous)', async () => {
    // Mock the hook to return playlist data
    (usePlaylistData as vi.Mock).mockReturnValue({
      data: [
        { id: 1, title: 'Painted in Blue', artist: 'Soul Canvas', duration: '5:55' },
        { id: 2, title: 'Tidal Drift', artist: 'Echoes of the Sea', duration: '8:02' },
      ],
      loading: false,
    });

    render(<MusicPlayer />);

    // Initially, the first song should be displayed
    expect(await screen.findByText('Painted in Blue')).toBeInTheDocument();

    // Simulate clicking the "Next" button
    fireEvent.click(screen.getByRole('button', { name: /next/i }));
    expect(await screen.findByText('Tidal Drift')).toBeInTheDocument();

    // Simulate clicking the "Previous" button
    fireEvent.click(screen.getByRole('button', { name: /previous/i }));
    expect(await screen.findByText('Painted in Blue')).toBeInTheDocument();
  });

  it('toggles shuffle mode', async () => {
    // Mock the hook to return playlist data
    (usePlaylistData as vi.Mock).mockReturnValue({
      data: [
        { id: 1, title: 'Painted in Blue', artist: 'Soul Canvas', duration: '5:55' },
        { id: 2, title: 'Tidal Drift', artist: 'Echoes of the Sea', duration: '8:02' },
      ],
      loading: false,
    });

    render(<MusicPlayer />);

    // Simulate clicking the "Shuffle" button
    const shuffleButton = screen.getByRole('button', { name: /shuffle/i });
    fireEvent.click(shuffleButton);

    // Assuming there's some visual change to indicate shuffle is enabled
    expect(shuffleButton).toHaveClass('enabled');
  });
});
