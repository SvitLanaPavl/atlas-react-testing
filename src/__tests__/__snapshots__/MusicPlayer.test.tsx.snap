// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MusicPlayer component > renders correctly 1`] = `
<div>
  <div
    class="bg-customPurple-25 mx-auto flex h-full w-full max-w-5xl flex-col overflow-hidden rounded-lg shadow-lg md:flex-row"
  >
    <div
      class="p-6 md:w-1/2 flex flex-col justify-between animate-pulse"
    >
      <div
        class="relative mb-6 aspect-square bg-customPurple-50 rounded-md"
      />
       
      <div
        class="mb-4 h-6 bg-customPurple-50 rounded w-3/4"
      />
       
      <div
        class="mb-2 h-4 bg-customPurple-50 rounded w-1/2"
      />
       
      <div
        class="flex items-center justify-between mb-5"
      >
        <div
          class="h-10 w-10 bg-customPurple-50 rounded"
        />
         
        <div
          class="h-10 w-10 bg-customPurple-50 rounded"
        />
        <div
          class="h-10 w-10 bg-customPurple-50 rounded"
        />
        <div
          class="h-10 w-10 bg-customPurple-50 rounded"
        />
        <div
          class="h-10 w-10 bg-customPurple-50 rounded"
        />
      </div>
      <div
        class="h-2 w-full bg-customPurple-50 rounded"
      />
       
    </div>
    <div
      class="space-y-4 p-6 md:w-1/2"
    >
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
      <div
        class="w-full animate-pulse flex justify-between"
      >
        <div
          class="flex-grow mr-4"
        >
          <div
            class="h-4 bg-customPurple-50 rounded w-3/4"
          />
           
          <div
            class="h-3 bg-customPurple-50 rounded w-1/2 mt-2"
          />
           
        </div>
        <div
          class="h-8 w-8 bg-customPurple-50 rounded flex-shrink-0"
        />
         
      </div>
    </div>
  </div>
</div>
`;
